benchmarks/Bio/BCR.jmd

```diff
@@ -14,138 +14,322 @@ Catalyst network model to ODEs.
+    ylimit = 10 .^ (floor(log10(minimum(times))), ceil(log10(maximum(times))))
+    return xlimit,ylimit
+end
+```
+
+#### Declare pre-conditioners
+```julia
+if false
+    using IncompleteLU, LinearAlgebra
+
+    jaccache = sparsejacprob.f.jac(oprob.u0,oprob.p,0.0)
+    W = I - 1.0*jaccache
+    prectmp = ilu(W, τ = 50.0)
+    preccache = Ref(prectmp)
+
+    τ1 = 1e2
+    function psetupilu(p, t, u, du, jok, jcurPtr, gamma)
+        if jok
+            sparsejacprob.f.jac(jaccache,u,p,t)
+            jcurPtr[] = true
+
+            # W = I - gamma*J
+            @. W = -gamma*jaccache
+            idxs = diagind(W)
+            @. @view(W[idxs]) = @view(W[idxs]) + 1
+
+            # Build preconditioner on W
+            preccache[] = ilu(W, τ = τ1)
+        end
+    end
+    function precilu(z,r,p,t,y,fy,gamma,delta,lr)
+        ldiv!(z,preccache[],r)
+    end
+
+    τ2 = 1e10
+    function incompletelu(W,du,u,p,t,newW,Plprev,Prprev,solverdata)
+        if newW === nothing || newW
+            Pl = ilu(convert(AbstractMatrix,W), τ = τ2)
+        else
+            Pl = Plprev
+        end
+        Pl,nothing
+    end;
+end
+```
+
+#### Sets tolerances
+
+```julia
+if false
+    abstols = 1.0 ./ 10.0 .^ (5:8)
+    reltols = 1.0 ./ 10.0 .^ (5:8);
+end
+```
+
+## Work-Precision Diagrams (CVODE and lsoda solvers)
+
+#### Declare solvers.
+
+```julia
+if false
+    setups = [
+            Dict(:alg=>lsoda(), :prob_choice => 1),
+            Dict(:alg=>CVODE_BDF(), :prob_choice => 1),
+            Dict(:alg=>CVODE_BDF(linear_solver=:LapackDense), :prob_choice => 1),
+            Dict(:alg=>CVODE_BDF(linear_solver=:GMRES), :prob_choice => 1),
+            Dict(:alg=>CVODE_BDF(linear_solver=:GMRES,prec=precilu,psetup=psetupilu,prec_side=1), :prob_choice => 2),
+            Dict(:alg=>CVODE_BDF(linear_solver=:KLU), :prob_choice => 3),
+            ];
+end
+```
+
+
+#### Plot Work-Precision Diagram.
+
+```julia
+if false
+    wp = WorkPrecisionSet([oprob,oprob_sparse,sparsejacprob],abstols,reltols,setups;error_estimate=:l2,
+                        saveat=tf/10000.,appxsol=[test_sol,test_sol,test_sol],maxiters=Int(1e9),numruns=2)
+
+    names = ["lsoda" "CVODE_BDF" "CVODE_BDF (LapackDense)" "CVODE_BDF (GMRES)" "CVODE_BDF (GMRES, iLU)" "CVODE_BDF (KLU, sparse jac)"]
+    xlimit,ylimit = plot_settings(wp)
+    plot(wp;label=names,xlimit=xlimit,ylimit=ylimit)
+end
+```
+
+## Work-Precision Diagrams (various Julia solvers)
+
+#### Declare solvers (using default linear solver).
+
+```julia
+if false
+    setups = [
+            Dict(:alg=>TRBDF2(autodiff=false)),
+            Dict(:alg=>QNDF(autodiff=false)),
+            Dict(:alg=>FBDF(autodiff=false)),
+            Dict(:alg=>KenCarp4(autodiff=false)),
+            Dict(:alg=>Rosenbrock23(autodiff=false)),
+            Dict(:alg=>Rodas4(autodiff=false)),
+            Dict(:alg=>Rodas5P(autodiff=false)),
+            Dict(:alg=>ROCK2(autodiff=false)), 
+            Dict(:alg=>ROCK4(autodiff=false))

```

Comment:
ROCK methods will never finish here. I don't understand what you're doing. This seems copy pasted from other benchmarks. Different benchmark problems are very different!

Reply:
Yes, I think I misunderstood earlier, wasn't sure where ROCK belong so did it for all in the initial run.